### PR TITLE
BRITECHSRM-396: Profile put method

### PR DESCRIPTION
**Pull request by jebbinSCL**
Add profile put method and refactorings

### CODE REVIEW

**Comment by pallenSL** [comment]
Not down to your changes, but the separation of concerns and error handling here are awful!  The repo should never return HttpStatus codes.
I'm also not a fan of the one-line "if error failwith" stuff - it is only hiding the complexity, not removing it.

File: Sessions/ProfilesRepository.fs [hunk 1]
```diff
@@ -40,49 +40,99 @@ let modelToEntity (model : Profile) : ProfileEntity =
...
         connection.Close()
-        Success profile
+
+        if box profileEntity |> isNull then
+            Failure { HttpStatus = HttpStatusCode.NotFound
+                      Message = sprintf "Profile with id %A does not exist" profileId }
+        else 
+            let profile = entityToModel profileEntity handleEntities
+            Success profile
     with
```
Comment (pallenSL) [inline Sessions/ProfilesRepository.fs]: The structure here would be far clearer as:
get profile
if profile not null
  get handle
  Success profile profile handle
else
  scream/log/throw

And seemingly all to Close the connection, which the use should do for us.

File: Sessions/ProfilesRepository.fs [hunk 2]
```diff
@@ -40,49 +40,99 @@ let modelToEntity (model : Profile) : ProfileEntity =
...
-        connection.Open()
-    
+        connection.Open()    
         use transaction = connection.BeginTransaction()
+
         let profileInsertCount = connection.Execute(@"insert profiles(id,forename,surname,rating,imageUrl) values (@Id,@Forename,@Surname,@Rating,@ImageUrl)", profileEntity)
-        if profileInsertCount <> 1 then failwith "Incorrect number of inserted profiles found. Profile insert failed"
-        let handleEntities = profile.Handles |> Seq.map (handleModelToEntity newId)
+        if profileInsertCount <> 1 then failwith "Profile insert failed"
+
```
Comment (pallenSL) [inline Sessions/ProfilesRepository.fs]: I think we should either be leaving these errors to a throw, or properly structuring our code.  Does connection.Execute() throw on error?

File: Sessions/ProfilesRepository.fs [hunk 3]
```diff
@@ -40,49 +40,99 @@ let modelToEntity (model : Profile) : ProfileEntity =
...
+            | Success _ ->
+                let profileEntity = modelToEntity profile
+                let handleEntities = profile.Handles |> Seq.map (handleModelToEntity profileEntity.Id)
+
+                use connection = getConnection()
+                connection.Open()
+                use transaction = connection.BeginTransaction()
+
+                let profileUpdateCount = connection.Execute(@"update profiles set forename=@Forename,surname=@Surname,rating=@Rating,imageurl=@ImageUrl,bio=@Bio where Id = @Id",profileEntity)
+                if profileUpdateCount <> 1 then failwith "Update of profile data failed"
```
Comment (pallenSL) [inline Sessions/ProfilesRepository.fs]: Again here - are we using exceptions, or code flow?

File: Sessions/ProfilesRepository.fs (hunk 2 above)
Comment (jebbinSCL) [inline Sessions/ProfilesRepository.fs]: As far as I can tell, there seem to be some failures that it won't error on. The documentation shows examples doing the same thing but its not clear why other than validation

**Comment by jebbinSCL** [comment]
Updated to deal with @pallenSL 's comments.

**Comment by pallenSL** [comment]
Much clearer!
:shipit:

File: Sessions/ProfilesController.fs [hunk 4]
```diff
@@ -6,19 +6,38 @@ open System.Net.Http
 open System.Web.Http
```
Comment (csscottlogic) [inline Sessions/ProfilesController.fs]: There's a lot of commonality between the methods in this file. It's OK for this PR, but it looks as if we need to factor out Success / Failure handling.

File: Sessions/ProfilesRepository.fs [hunk 5]
```diff
@@ -40,49 +40,91 @@ let modelToEntity (model : Profile) : ProfileEntity =
...
         Log.Error("addProfile() - Exception: {0}", ex)
         Failure { HttpStatus = HttpStatusCode.BadRequest
                   Message = ex.Message }
 
+let updateProfile (pid: Guid) (profile : Profile) = 
+    if pid <> profile.Id then 
+        Failure { HttpStatus = HttpStatusCode.BadRequest; Message = "Invalid Data. specified profile Id in request url does not match Id of input profile" } 
+    else 
+
+    let handleDoesNotExistIn handles handle =
```
Comment (csscottlogic) [inline Sessions/ProfilesRepository.fs]: Should this not be indented?

File: Sessions/ProfilesRepository.fs [hunk 6]
```diff
@@ -40,49 +40,91 @@ let modelToEntity (model : Profile) : ProfileEntity =
...
+    if pid <> profile.Id then 
+        Failure { HttpStatus = HttpStatusCode.BadRequest; Message = "Invalid Data. specified profile Id in request url does not match Id of input profile" } 
+    else 
+
+    let handleDoesNotExistIn handles handle = 
+        handles |> Seq.exists (fun hl -> hl.Type = handle.Type && hl.Identifier = handle.Identifier) |> not
+    try 
+        match getProfile pid with
+        | Success _ ->
+            let profileEntity = modelToEntity profile
```
Comment (csscottlogic) [inline Sessions/ProfilesRepository.fs]: This should be factored out

**Comment by csscottlogic** [comment]
My overall impression here is that there is quite a bit of commonality in the methods in ```ProfilesController``` and that the code in ```ProfilesRepository``` is too verbose and needs factoring.

**Comment by csscottlogic** [comment]
I'm happy to :shipit: to keep things moving along, but if it were me I'd address the points above

File: Sessions/ProfilesController.fs (hunk 4 above)
Comment (jebbinSCL) [inline Sessions/ProfilesController.fs]: Looks like a computational expression might be the best fit for this. Three methods are very similar and have the same workflow, but the individual steps all have a different number of arguments. Do we want to approach this yet?

Comment (csscottlogic) [inline Sessions/ProfilesController.fs]: Let's get the feature finished and then discuss this.

**Comment by jebbinSCL** [comment]
I've refactored the methods in ProfilesRepository. I'm thinking about splitting the handles methods from the profiles, but they are a bit tied together. I think its good to leave for now and deal with later if you are OK with that.

File: Sessions/ProfilesRepository.fs [hunk 7]
```diff
@@ -40,79 +40,143 @@ let modelToEntity (model : Profile) : ProfileEntity =
...
+                match insertHandleEntities connection newHandles with
+                | Success () -> 
+                    match deleteHandleEntities connection handlesToDelete with
+                    | Success () -> 
+                        transaction.Commit()
+                        Success ()
+                    | failure -> failure
+                | failure -> failure
+            | failure -> failure
+        | Failure error ->
```
Comment (csscottlogic) [inline Sessions/ProfilesRepository.fs]: I'm not sure that "failure" repeated 4 times with different levels of indentation is that readable?
This looks like a computation expression that implements ROP (rails oriented programming), i.e. run these statements in order and stop on the first one that fails.
We can merge for now, but this shouldn't be what we settle for.

**Pull request merged**